RecoJets/JetProducers/plugins/VirtualJetProducer.cc

```diff
@@ -846,6 +856,114 @@ void VirtualJetProducer::writeCompoundJets(  edm::Event & iEvent, edm::EventSetu
 
 }
 
+/// function template to write out the outputs
+template< class T>
+void VirtualJetProducer::writeJetsWithConstituents(  edm::Event & iEvent, edm::EventSetup const& iSetup)
+{
+  if ( verbosity_ >= 1 ) {
+    std::cout << "<VirtualJetProducer::writeJetsWithConstituents (moduleLabel = " << moduleLabel_ << ")>:" << std::endl;
+  }
+
+  // get a list of output jets  MV: make this compatible with template
+  auto jetCollection = std::make_unique<reco::PFJetCollection>();
+  
+  // this is the mapping of jet to constituents
+  std::vector< std::vector<int> > indices;
+  // this is the list of jet 4-momenta
+  std::vector<math::XYZTLorentzVector> p4_Jets;
+  // this is the jet areas
+  std::vector<double> area_Jets;
+
+  // get a list of output constituents
+  auto constituentCollection = std::make_unique<reco::PFCandidateCollection>();
+  
+  // This will store the handle for the constituents after we write them
+  edm::OrphanHandle<reco::PFCandidateCollection> constituentHandleAfterPut;
+    
+  // Loop over the jets and extract constituents
+  std::vector<fastjet::PseudoJet> constituentsSub;
+  std::vector<fastjet::PseudoJet>::const_iterator it = fjJets_.begin(),
+    iEnd = fjJets_.end(),
+    iBegin = fjJets_.begin();
+  indices.resize( fjJets_.size() );
+
+  for ( ; it != iEnd; ++it ) {
+    fastjet::PseudoJet const & localJet = *it;
+    unsigned int jetIndex = it - iBegin;
+    // Get the 4-vector for the hard jet
+    p4_Jets.push_back( math::XYZTLorentzVector(localJet.px(), localJet.py(), localJet.pz(), localJet.e() ));
+    double localJetArea = 0.0;
+    if ( doAreaFastjet_ && localJet.has_area() ) {
+      localJetArea = localJet.area();
+    }
+    area_Jets.push_back( localJetArea );
+
+    // create the constituent list
+    std::vector<fastjet::PseudoJet> constituents,ghosts;
+    if ( it->has_pieces() )
+      constituents = it->pieces();
+    else if ( it->has_constituents() )
+      fastjet::SelectorIsPureGhost().sift(it->constituents(), ghosts, constituents); //filter out ghosts
+
+    //loop over constituents of jet (can be subjets or normal constituents)
+    for (fastjet::PseudoJet const& constit : constituents) {
```

Comment:
please move the if outside the for loop (better if you could use ifdefs instead for verbosity printouts.